crates/sui-core/src/consensus_adapter.rs

```diff
@@ -170,18 +180,127 @@ impl SubmitToConsensus for TransactionsClient<sui_network::tonic::transport::Cha
     }
 }
 
+pub struct ConnectionMonitorStatusForTests {}
+
+impl CheckConnection for ConnectionMonitorStatusForTests {
+    fn check_connection(&self, _authority: &AuthorityName) -> Option<ConnectionStatus> {
+        Some(ConnectionStatus::Connected)
+    }
+}
+
+pub trait CheckConnection {
+    fn check_connection(&self, authority: &AuthorityName) -> Option<ConnectionStatus>;
+}
+
+pub struct ConnectionMonitorStatus {
+    /// Current connection statuses forwarded from the connection monitor
+    pub connection_statuses: Arc<DashMap<AuthorityName, ConnectionStatus>>,
+    /// Join handle for the Connection Monitor Listener
+    pub connection_monitor_listener_handle: JoinHandle<()>,
+}
+
+pub struct ConnectionMonitorListener {
+    /// Receiver from the Connection Monitor
+    receiver: Receiver<(PeerId, ConnectionStatus)>,
+    /// Map to look up the connection statuses as a latency aid during submit to consensus
+    pub current_connection_statuses: Arc<DashMap<AuthorityName, ConnectionStatus>>,
+    /// Map to populate the current connection statuses from the receiver
+    peer_id_to_authority_names: HashMap<PeerId, AuthorityName>,
+}
+
+impl CheckConnection for ConnectionMonitorStatus {
+    fn check_connection(&self, authority: &AuthorityName) -> Option<ConnectionStatus> {
+        let res = match self.connection_statuses.try_get(authority) {
+            TryResult::Present(c) => Some(c.value().clone()),
+            TryResult::Absent => None,
+            TryResult::Locked => {
+                // update is in progress, assume the status is still or becoming disconnected
+                Some(ConnectionStatus::Disconnected)
+            }
+        };
+        res
+    }
+}
+
+impl ConnectionMonitorListener {
+    pub async fn spawn(
+        receiver: Receiver<(PeerId, ConnectionStatus)>,
+        peer_id_to_authority_names: HashMap<PeerId, AuthorityName>,
+    ) -> ConnectionMonitorStatus {
+        let connection_statuses = DashMap::with_capacity(peer_id_to_authority_names.len());
+        for (_, authority_name) in peer_id_to_authority_names.iter() {
+            // initialize all to connected as default,if we don't have a consensus monitor running so
+            // the fallback behavior is we submit to consensus once and not 2f+1 times
+            connection_statuses.insert(*authority_name, ConnectionStatus::Connected);
+        }
+        let current_connection_statuses = Arc::new(connection_statuses);
+
+        debug!(" spawning connection monitor listener");
+        ConnectionMonitorStatus {
+            connection_statuses: current_connection_statuses.clone(),
+            connection_monitor_listener_handle: spawn_monitored_task!(Self {
+                receiver,
+                current_connection_statuses,
+                peer_id_to_authority_names
+            }
+            .run()),
+        }
+    }
+
+    async fn run(&mut self) {
```

Comment:
We need some story to understand when does this task finish (to make sure it actually completes on epoch boundary).
Is it guaranteed that `self.receiver` is dropped when epoch ends? Why? 
(Can you reply to this question as a comment to the `run` method as this is worth mentioning in the code itself?)